src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/script.js.template

```diff
@@ -0,0 +1,481 @@
+                tab.setAttribute('tabindex',0);
+            } else {
+                tab.className = tableTab;
+                tab.setAttribute('aria-selected', false);
+                tab.setAttribute('tabindex',-1);
+            }
+        });
+}
+
+function switchTab(e) {
+    var selected = document.querySelector('[aria-selected=true]');
+    if (selected) {
+        if ((e.keyCode === 37 || e.keyCode === 38) && selected.previousSibling) {
+            // left or up arrow key pressed: move focus to previous tab
+            selected.previousSibling.click();
+            selected.previousSibling.focus();
+            e.preventDefault();
+        } else if ((e.keyCode === 39 || e.keyCode === 40) && selected.nextSibling) {
+            // right or down arrow key pressed: move focus to next tab
+            selected.nextSibling.click();
+            selected.nextSibling.focus();
+            e.preventDefault();
+        }
+    }
+}
+
+var updateSearchResults = function() {};
+
+function indexFilesLoaded() {
+    return moduleSearchIndex
+        && packageSearchIndex
+        && typeSearchIndex
+        && memberSearchIndex
+        && tagSearchIndex;
+}
+// Copy the contents of the local snippet to the clipboard
+function copySnippet(button) {
+    copyToClipboard(button.nextElementSibling.innerText);
+    switchCopyLabel(button, button.firstElementChild);
+}
+function copyToClipboard(content) {
+    var textarea = document.createElement("textarea");
+    textarea.style.height = 0;
+    document.body.appendChild(textarea);
+    textarea.value = content;
+    textarea.select();
+    document.execCommand("copy");
+    document.body.removeChild(textarea);
+}
+function switchCopyLabel(button, span) {
+    var copied = span.getAttribute("data-copied");
+    button.classList.add("visible");
+    var initialLabel = span.innerHTML;
+    span.innerHTML = copied;
+    setTimeout(function() {
+        button.classList.remove("visible");
+        setTimeout(function() {
+            if (initialLabel !== copied) {
+                span.innerHTML = initialLabel;
+            }
+        }, 100);
+    }, 1900);
+}
+function setTopMargin() {
+    // Dynamically set scroll margin to accomodate for draft header
+    var headerHeight = Math.ceil(document.querySelector("header").offsetHeight);
+    document.querySelector(":root")
+        .style.setProperty("--nav-height", headerHeight + "px");
+}
+document.addEventListener("readystatechange", (e) => {
+    if (document.readyState === "interactive") {
+        setTopMargin();
+    }
+    if (sessionStorage.getItem("sidebar") === "hidden") {
+        const sidebar = document.querySelector(".main-grid nav.toc");
+        if (sidebar) sidebar.classList.add("hide-sidebar");
+    }
+});
+document.addEventListener("DOMContentLoaded", function(e) {
+    setTopMargin();
+    // Make sure current element is visible in breadcrumb navigation on small displays
+    const subnav = document.querySelector("ol.sub-nav-list");
```

Comment:
This is but one instance of many similar `querySelector` calls that leverage CSS class names and/or HTML ids.

In the Java code, we used named constants for such names, defined in `HtmlStyle` and `HtmlId`, which allows us to use an IDE to track their usage. But that does not work across the language boundary into code like this. I wonder (just asking) if it is worth manually marking the Java constants in some way to indicate the names are also used in JavaScript code.

Reply:
Interesting suggestion! What would the advantages of using annotations be compared to just using comments, which could also contain a description of where/how the constants are used?

Reply:
Annotations may provide slightly more checking than stylized comments ... such as for typos and/or renames, etc.  They can also be the basis for checks in tests.  For example, the `javac` code (and to a lesser extent, `javadoc`) uses the `@DefinedBy` annotation to indicate which methods are exported through public API, and should not be modified without due consideration. IIRC, there are tests that methods that have such an annotation are indeed part of the stated API (given as an argument to the annotation.)